### PR TITLE
Protect villages from being overbuilt

### LINKED ISSUE #1075: Builder should break entities when upgrading.
Hey, my Builder dont want to break Item Frames when upgrading a Hut
Hes done with lvl 5 and now there are 2 item frames on the wall who arent in the schematic from the lvl 5 hut

### LINKED ISSUE #1116: [request] prevent villages from be captured by townhalls
Would be awesome if you could add a feature where when a town hall is placed such that encompasses a village, that the village remains public access. Better yet, the town hall refuses to place because it overlaps the village.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #1116
Closes #1075

# Changes proposed in this pull request:
- Doesn't allow players to build over colonies if the config option is turned on.
- Only works after the first colony of the world has been placed already. (Doesn't make sense in sp anyway)
- Fixes also the problem where entities are not dropped when the building gets upgraded and they're not needed anymore.
- Also fixes problems with entities in 1.11.
Review please

### CODE REVIEW

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 34 issues
* ![MINOR][MINOR] 34 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [blocks](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/blocks): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [gui](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/client/gui): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [render](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/client/render): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [colony](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/colony): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [buildings](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/colony/buildings): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [permissions](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/colony/permissions): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [requestsystem](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/colony/requestsystem): Add a 'package-info.java' file to document the 'requestsystem' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [init](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/colony/requestsystem/init): Add a 'package-info.java' file to document the 'init' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [locations](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/colony/requestsystem/locations): Add a 'package-info.java' file to document the 'locations' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [requests](https://github.com/Minecolonies/minecolonies/blob/e2b59696c53c5968d0a7b6e3ea1b53bf24c119f7/src/main/java/com/minecolonies/coremod/colony/requestsystem/requests): Add a 'package-info.java' file to document the 'requests' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 45 issues
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 36 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L59](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L59): class "AbstractEntityAIStructure" has 39 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L792](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L792): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L52](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L52): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L177](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L177): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L190](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L190): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [Structure.java#L69](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L69): Reduce this class from 56 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PlacementHandlers.java#L72](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java#L72): Reduce this class from 31 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PlacementHandlers.java#L240](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java#L240): Reduce this class from 31 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PlacementHandlers.java#L441](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java#L441): Reduce this class from 36 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MINOR][MINOR] [blocks](https://github.com/Minecolonies/minecolonies/blob/4b1e03be11418cc25490f12dcda5d13421c20bdf/src/main/java/com/minecolonies/coremod/blocks): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — APPROVED** [review]

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 1]
```diff
@@ -435,11 +445,17 @@ public boolean isCorrectMD5(final String otherMD5)
...
      * Reads a template from an inputstream.
      */
-    private static Template readTemplateFromStream(final InputStream stream) throws IOException
+    private static Template readTemplateFromStream(final InputStream stream, final DataFixer fixer) throws IOException
     {
         final NBTTagCompound nbttagcompound = CompressedStreamTools.readCompressed(stream);
-        final Template template = new Template();
-        template.read(nbttagcompound);
+
+        if (!nbttagcompound.hasKey("DataVersion", 99))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java]: MAJOR Assign this magic number 99 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 2]
```diff
@@ -435,11 +445,17 @@ public boolean isCorrectMD5(final String otherMD5)
...
+        {
+            nbttagcompound.setInteger("DataVersion", 500);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java]: MAJOR Assign this magic number 500 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 51 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![MAJOR][MAJOR] 12 major
* ![MINOR][MINOR] 36 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [Structure.java#L126](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/structures/helpers/Structure.java#L126): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L235](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/structures/helpers/Structure.java#L235): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L248](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/structures/helpers/Structure.java#L248): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L60](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L60): class "AbstractEntityAIStructure" has 39 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L804](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L804): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L52](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L52): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L177](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L177): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L190](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L190): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [Structure.java#L71](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L71): Reduce this class from 56 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PlacementHandlers.java#L72](https://github.com/Minecolonies/minecolonies/blob/5d95e4df49a6c40a88bf1d37f49ef6b688f3f5d2/src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java#L72): Reduce this class from 31 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 3]
```diff
@@ -435,11 +445,17 @@ public boolean isCorrectMD5(final String otherMD5)
     /**
      * Reads a template from an inputstream.
      */
-    private static Template readTemplateFromStream(final InputStream stream) throws IOException
+    private static Template readTemplateFromStream(final InputStream stream, final DataFixer fixer) throws IOException
     {
         final NBTTagCompound nbttagcompound = CompressedStreamTools.readCompressed(stream);
+
+        if (!nbttagcompound.hasKey("DataVersion", 99))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:452]: MAJOR Assign this magic number 99 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 4]
```diff
@@ -435,11 +445,17 @@ public boolean isCorrectMD5(final String otherMD5)
...
+        {
+            nbttagcompound.setInteger("DataVersion", 500);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:454]: MAJOR Assign this magic number 500 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 51 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![MAJOR][MAJOR] 12 major
* ![MINOR][MINOR] 36 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [Structure.java#L126](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/structures/helpers/Structure.java#L126): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L235](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/structures/helpers/Structure.java#L235): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L248](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/structures/helpers/Structure.java#L248): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L60](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L60): class "AbstractEntityAIStructure" has 39 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L804](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L804): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L52](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L52): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L177](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L177): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L190](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L190): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [Structure.java#L71](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L71): Reduce this class from 56 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PlacementHandlers.java#L72](https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java#L72): Reduce this class from 31 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Asherslab** [comment]
can we put a sonar suppression on:
https://github.com/Minecolonies/minecolonies/blob/311222132d6b50e0df58f72f781a3ef1c9a226c5/src/main/java/com/minecolonies/structures/helpers/Structure.java#L126 (rule "squid:S2095") 
since it's blocking all PRs because of it? (Or if possible fix it xD)

**Comment by marchermans** [comment]
No we should not. This is not up for discussion either close the stream or make it autoclosable. If it is already autoclosable then i have said nothing, but the normal input stream is not as far as i am aware of.

So load the data and then close it, or make it autoclosable.

**Comment by Asherslab** [comment]
hmmm, this seems to be coming up in all the team city builds, not sure where it's being caused though.
`com.google.gson.stream.MalformedJsonException: Unterminated string at line 195 column 1`


#BlameMarvin

**Comment by marchermans** [comment]
I have no idea. I tried several Things, it is running on one of Marvins machines which i do not have acces to. So i have no idea what is going on here.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 51 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![MAJOR][MAJOR] 12 major
* ![MINOR][MINOR] 36 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [Structure.java#L126](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/structures/helpers/Structure.java#L126): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L235](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/structures/helpers/Structure.java#L235): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![BLOCKER][BLOCKER] [Structure.java#L248](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/structures/helpers/Structure.java#L248): Close this "InputStream" in a "finally" clause. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L60](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L60): class "AbstractEntityAIStructure" has 39 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIStructure.java#L804](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L804): Refactor this method to reduce its Cognitive Complexity from 32 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L52](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L52): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 31 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L177](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L177): The Cyclomatic Complexity of this method "requestMaterials" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [EntityAIStructureBuilder.java#L190](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java#L190): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [Structure.java#L71](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/coremod/entity/ai/util/Structure.java#L71): Reduce this class from 56 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [PlacementHandlers.java#L72](https://github.com/Minecolonies/minecolonies/blob/4d587b2d9bc813358c8acf623cbe02d64254f156/src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java#L72): Reduce this class from 31 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Asherslab** [comment]
still need to close that stream

**Comment by Raycoms** [comment]
Like I said, it has nothing to do with my pr. That's there for a year now.

2017-08-31 6:33 GMT+01:00 Asherslab <notifications@github.com>:

**Review by marchermans — APPROVED** [review]
@Asherslab I took a look at the Situation, and overruled sonar on that one, it does not seem to get the Autoclosable Stream System that MC uses for some resources and files.

Code changes as such are accepted.

**Comment by Asherslab** [comment]
Okay, sounds good. Everything else looks awesome! :shipit:

**Pull request merged**